crates/starknet-devnet-core/src/starknet/estimations.rs

```diff
@@ -132,14 +131,24 @@ fn estimate_transaction_fee<S: StateReader>(
     let transaction_execution_info = transaction.execute(transactional_state, block_context)?;
 
     // reverted transactions can only be Invoke transactions
-    if let (true, Some(revert_error)) =
-        (return_error_on_reverted_execution, transaction_execution_info.revert_error)
-    {
-        // TODO Users would probably prefer a structured error, but according to the RPC spec, a
-        // string is allowed. We should improve this.
-        return Err(Error::ContractExecutionError(ErrorStack::from_str_err(
-            &revert_error.to_string(),
-        )));
+    match transaction_execution_info.revert_error {
+        Some(revert_error) if return_error_on_reverted_execution => {
+            // TODO Users would probably prefer a structured error, but according to the RPC spec, a
+            // string is allowed. We should improve this for non Execution error
+            let revert_error_string = revert_error.to_string();
+            match revert_error {
+                blockifier::transaction::objects::RevertError::Execution(error_stack) => {
+                    return Err(Error::ContractExecutionError(ContractExecutionError::from((
+                        error_stack,
+                        revert_error_string,
+                    ))));
```

Comment:
We can probably remove the TODO comment. Also, IMO the `revert_error_string` parameter is unneeded. If really needed, `error_stack` can be stringified inside the `From` converter.

Reply:
Addressed in https://github.com/0xSpaceShard/starknet-devnet/pull/722/commits/ee23c23dbbee0e5117193691398ce5720e7facad